Write deprecation logs to a data stream

pugnascotia opened the pull request:
Closes #46106. Introduce a mechanism for writing deprecation logs to a data stream as well as to disk.

Test by running `./gradlew run` and then:

```
curl -u elastic-admin:elastic-password http://localhost:9200/_flush/synced?pretty
curl -u elastic-admin:elastic-password http://localhost:9200/logs-deprecation-elasticsearch/_search?pretty
```

This implementation reworks deprecation logging to rely on log4j by introducing filters and appenders, and a custom log level. The existing `deprecation` X-Pack plugin simply configures an extra appender to capture messages to an index. A "component" class manages most of the interface between the cluster and the appender.

This approach can be extended for other types of logging, e.g. security by introducing further custom log levels.

pugnascotia committed:
Introduce DeprecationIndexingService

pugnascotia committed:
First go at ensuring the data stream template is loaded

pugnascotia committed:
Tweaks index name

pugnascotia committed:
WIP - Adding tests for DeprecationIndexingService

pugnascotia committed:
More tests

pugnascotia committed:
Fix template index pattern

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Logging)

pugnascotia commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -45,8 +46,8 @@
-    private DeprecationLogger(Logger parentLogger) {
-        deprecationLogger = new ThrottlingAndHeaderWarningLogger(parentLogger);
+    public DeprecationLogger(Logger parentLogger) {
+        deprecationLogger = new ThrottlingAndHeaderWarningLogger(deprecatedLoggerName(parentLogger.getName()), indexingService);
```
Do I need to be more defensive about how `indexingService` is set and supplied to the logger here?

pgomulka replied:
some assertion might indeed be helpful. It would prevent some accidental refactoring failures - people might not notice until some deprecation log is actually triggered.
I would also expect soem deprecation logger test to fail now.

pugnascotia commented:
I should add, the set of indexed fields in this PR may need to evolve, as it becomes clearer what we need to support the Upgrade Assistant. It should be a reasonable starting point however.

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

jakelandis commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,154 @@
+    /*
+     * Attempts to load a template for the deprecation logs data stream
+     */
+    private void loadTemplate() {
```
This duplicates some recent functionality in https://github.com/elastic/elasticsearch/tree/master/x-pack/plugin/stack and I believe is the only default template installed outside of x-pack, we should probably discuss where this functionality should live. cc @dakrone

dakrone replied:
Yeah, we have a whole tool used for installing templates and policies that should be used, it's currently in xpack, but we could move it to core if needed for this (I haven't looked at the rest of this PR, just a drive-by comment)

jakelandis commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,154 @@
+            payload.put("params", params);
+        }
+
+        new IndexRequestBuilder(client, IndexAction.INSTANCE).setIndex(DATA_STREAM_NAME)
```
I would suggest to use the BulkProcessor (https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java) since it allows for a configurable flush interval and optional incremental backoff retry logic.

jakelandis commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,154 @@
+
+    private static final String DEPRECATION_ORIGIN = "deprecation";
+
+    public static final Setting<Boolean> WRITE_DEPRECATION_LOGS_TO_INDEX = Setting.boolSetting(
```
Should the setting allow to flex between writing to a log, an index, or both ?

pugnascotia commented:
Moving to draft status while we reconsider some things.

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,154 @@
+            return;
+        }
+
+        Map<String, Object> payload = new HashMap<>();
```
we discussed this offline, but I will leave the comment for further discussion. Should we unify the format/fields generated by deprecation logs with what will be indexed?
Should we expect that ingesting 'manually' deprecation logs would give the same result (same format/fields in deprecation index) as when indexing from here?

I wonder if using the layout we use for formatting deprecation logs to ECS
```
SLogMessage deprecationMessage = DeprecatedMessage.of("x", "ss {}", 1);
LogEvent event = Log4jLogEvent.newBuilder()
    .setMessage(deprecationMessage)
    .setLoggerName("logger_name")
    .build();
final LoggerContext context = (LoggerContext) LogManager.getContext(false);
Layout<? extends Serializable> deprecated = context.getConfiguration().getAppender("deprecated").getLayout();
deprecated.toSerializable(event)
```

I am happy with simplest first approach, but might be worth considering this for the future

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,154 @@
+        Map<String, Object> payload = new HashMap<>();
+        payload.put("@timestamp", Instant.now().toString());
+        payload.put("key", key);
+        payload.put("message", message);
```
message and params I think should be 'expanded' otherwise `deprecationLogger.deprecate("msg {}", 123) ` would create two fields `"message: " msg" , "params" : 123` instead of just `msg 123`
or we could think of using log4j layout as per previous comment

pugnascotia replied:
I actually did that intentionally, because I thought it might be more useful...for some reason. Possibly that was before I changed the code to always require a key, so maybe now it makes sense to expand the parameters?

pgomulka replied:
I think it would be better to expand them before indexing. 
Re `key` field I don't think we need it. We do not log `key` in deprecation json (nor plaintext before) .
btw Would it be far more expensive to index every deprecation log message (no rate limiting)?

pugnascotia committed:
Rework deprecation code so that log handlers are more generic

pugnascotia committed:
Move DeprecationIndexingService to x-pack depreacation plugin

pugnascotia committed:
Don't need a custom template any more :tada:

pugnascotia committed:
Remove unused import

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
License header fix

pugnascotia commented:
I've reworked the implementation to move it to X-Pack, and stripped out all the template stuff completely, as X-Pack provides everything we need.

pugnascotia committed:
Remove unused imports

pugnascotia committed:
Add Javadoc

pugnascotia committed:
Silence log checker error

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -23,30 +23,43 @@
+    /** Log handlers are are not tied to this specific <code>DeprecationLogger</code> instance */
+    private static final List<DeprecatedLogHandler> additionalHandlers = new ArrayList<>();
+
+    /** Log handlers are are tied to this specific <code>DeprecationLogger</code> instance */
```
nit: typo doubled `are` ?

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -23,30 +23,43 @@
 public class DeprecationLogger {
-    private final ThrottlingAndHeaderWarningLogger deprecationLogger;
+    /** Log handlers are are not tied to this specific <code>DeprecationLogger</code> instance */
+    private static final List<DeprecatedLogHandler> additionalHandlers = new ArrayList<>();
```
will adding/remove handlers be done in another thread? Or do we expect this to be set-up before any logging is allowed? (pre logconfig is loaded)
I wonder if this should be thread safe.
Also why not keeping them both together in the same collection?. Looks like the iteration in `withDeprecation` is the same (at least at the moment)

pgomulka commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -93,10 +114,16 @@ public DeprecationLoggerBuilder deprecate(final String key, final String msg, fi
-            deprecationLogger.throttleLogAndAddWarning(key, deprecationMessage);
+            rateLimiter.limit(opaqueId + key, () -> {
+                ESLogMessage deprecationMessage = DeprecatedMessage.of(opaqueId, msg, params);
+                for (DeprecatedLogHandler handler : handlers) {
```
as per comment above, I wonder if this could be single collection

pugnascotia committed:
Subtitute log message params, simplify things

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
Fixes

pugnascotia commented:
@pgomulka I've reworked the code a bit. In particular, I realised that by pulling up the rate-limiting to `DeprecationLogger`, it meant that the `HeaderWarning.addWarning(...)` calls were also being limited, which wasn't correct. After moving things around a bit, I ended up only keeping a collection for handlers added externally.

I also defaulted the new setting to `false`, since the YAML tests don't automatically clean up data streams (yet).

Regarding the `key`, I wanted to keep that for now because I think it will be useful when we update the Migration Assistant in Kibana. We can remove `key` if we find that it isn't useful.

I don't think we want to index every single deprecation message. If a user enabled deprecation indexing on a busy cluster using a lot of deprecated features, we could be writing an awful lot of documents for no obvious benefit other than giving a overal count of deprecation warnings. As I understand it, from an upgrade perspective 1 deprecation is the same as 100 - a user has to address it eventually to upgrade.

pgomulka commented on x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,128 @@
+     *                     Useful when aggregating the recorded messages.
+     * @param esLogMessage the message to log
+     */
+    public void log(String key, String xOpaqueId, ESLogMessage esLogMessage) {
```
Do we expect this to be called from a different context than a deprecation logs?

if `esLogMessage` is a deprecation then it should contain xOpaqueId - so probably not needed to have that argument.
I have refactored `DeprecationMessage` not to inherit from `ESLogMessage` a while back, but maybe it should after all.. https://github.com/elastic/elasticsearch/pull/46702/files#r332013037 At the time I simply lacked the use cases to justify hierarchies for `IndexingSlowlogMessage`, `SearchSlowlog` etc

pugnascotia replied:
I removed the `xOpaqueId` argument, it works great to fetch it from `esLogMessage`.

pgomulka commented on x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,128 @@
+
+        String messagePattern = esLogMessage.getMessagePattern();
+        Object[] arguments = esLogMessage.getArguments();
+        String message = LoggerMessageFormat.format(messagePattern, arguments);
```
not the biggest fan of that `LoggerMessageFormat` . It duplicates what log4j can do for us now - I am not familiar with history behind this though. At some point I think we should refactor that logic to use log4j.
Happy with this for now.

How about logger name or node/cluster name fields?

pugnascotia committed:
Use BulkProcessor in DeprecationIndexingService

In order to make the service easily testable, the details of the
processor and thread pool etc are bundled up into a Consumer, which
hides it from the service.

pgomulka commented on x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,128 @@
+        payload.put("key", key);
+        payload.put("message", message);
+
+        if (isNullOrEmpty(xOpaqueId) == false) {
```
as per comment above. Possibly we can use xOpaqueId already provided on a message? WDYT?

pgomulka reviewed (commented):
looking good, I left comments.
Also a more general question. What is the ordering of using `DeprecationLogger` vs `DeprecationIndexingService`?
`DeprecationLogger` is create very early, but should not be used until logging context is configured - we will fail to start if there are any logs generated by `StatusLogger` (pre context loaded logger). But after that logger can be used. 
In other words - can this happen that `DeprecationLogger` is used but `DeprecationIndexingService` is not ready yet? Is it ok if we miss this logs being indexed?

pgomulka commented:
> In particular, I realised that by pulling up the rate-limiting to DeprecationLogger, it meant that the HeaderWarning.addWarning(...) calls were also being limited, which wasn't correct.

@pugnascotia I wonder if this would not be a good test case too add. We have some more 'high level' tests in `EvilLoggerTests` and `JsonLoggerTests`

pugnascotia committed:
Use the opaque ID from the EsLogMessage

pugnascotia commented on server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -23,30 +23,39 @@
 public class DeprecationLogger {
-    private final ThrottlingAndHeaderWarningLogger deprecationLogger;
+    /** Log handlers are not tied to this specific <code>DeprecationLogger</code> instance */
+    private static final List<DeprecatedLogHandler> additionalHandlers = Collections.synchronizedList(new ArrayList<>());
```
I can't reference `DeprecationIndexingService` directly because it only exists in X-Pack now. I don't expect multiple handlers, but I used a collection because it felt a bit wrong to have a mechanism for generic handlers but only support one.

Regarding synchronisation, I don't have strong opinions on the implementation since realistically, we'll either never modify the collection or only add a single handler.

rjernst reviewed (commented):
I want to suggest an alternative implementation. The main issue I see with this current PR is these additional handlers manipulated statically. They are essentially an extensibility point without a proper api.

I wonder if we could handle this by doing the following:
* Create a custom log level, DEPRECATION, in core
* DeprecationLogger becomes a simple wrapper over writing messages with that log level
* The logger we have for deprecation logger in core moves to a module only in oss distro
* The xpack implementation stays in xpack, essentially replacing the oss one that writes to a logfile

This is a very high level suggestion, so I'm sure their are holes in it, but the main point is we can use both log4j configuration that differs between oss and default, along with separate oss vs default implementations, to control the backend, rather than having an implicit, static extension point that is tricky to test.

pugnascotia committed:
Record cluster and node IDs

pugnascotia committed:
Tweaks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pgomulka commented on x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DeprecationIndexingService.java:
```diff
@@ -0,0 +1,107 @@
+    public DeprecationIndexingService(ClusterService clusterService, Consumer<IndexRequest> requestConsumer) {
+        this.requestConsumer = requestConsumer;
+
+        clusterService.addListener(this);
```
leaking this reference is dangerous - we have discussed this in the past and there is a meta issue to track and remove cluster state registrations like this. meta issue : https://github.com/elastic/elasticsearch/issues/38560
issue with explanation https://github.com/elastic/elasticsearch/issues/37861
can we try to refactor this ?

pugnascotia committed:
WIP - trying a log4j implementation

pugnascotia committed:
Reimplement deprecation logging using log4j

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-log4j

pgomulka commented:
> I want to suggest an alternative implementation. The main issue I see with this current PR is these additional handlers manipulated statically. They are essentially an extensibility point without a proper api.

how about the approach with SPI? something similar to what we with `NamedXContentProvider`. DeprecationLogger would still use a list, but would load implementations of `DeprecationService` upon creation. The default one would do nothing (or would log to a file?) and the `xpack` impl would log to an index.
WDYT?

pugnascotia commented:
I've been reworking the code to just use log4j abstractions, and it's pretty clean. I can log at a custom log level, then log4j takes care of all the plubming. I'll update the PR this week.

pugnascotia committed:
Reimplement using log4j abstractions

pugnascotia committed:
Polishing

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-log4j

pugnascotia committed:
More polishing

pugnascotia committed:
Include key in DeprecatedMessage.of(...)

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-log4j

pugnascotia committed:
Strip out UnionFilter

pugnascotia committed:
Build deprecation msg docs using EcsJsonLayout

pgomulka commented on distribution/src/config/log4j2.properties:
```diff
@@ -63,18 +63,23 @@ appender.deprecation_rolling.name = deprecation_rolling
 appender.deprecation_rolling.policies.size.size = 1GB
 appender.deprecation_rolling.strategy.type = DefaultRolloverStrategy
 appender.deprecation_rolling.strategy.max = 4
+
```
generic log4j2.properties comment
we probably need to update configs for docker too

pugnascotia replied:
I've updated the Docker configs too.

pgomulka reviewed (commented):
looking really nice. I really liked the filter approach.
left some comments

pugnascotia committed:
Checkstyle

pugnascotia committed:
Update docker log4j configs

pugnascotia committed:
Fix libs tests

pugnascotia committed:
Configure HeaderWarningAppender in ESTestCase

pugnascotia committed:
Fix docker log4j deprecation levels

pugnascotia commented on server/src/main/java/org/elasticsearch/common/logging/HeaderWarningAppender.java:
```diff
@@ -0,0 +1,63 @@
+import org.apache.logging.log4j.core.config.plugins.PluginFactory;
+import org.apache.logging.log4j.message.Message;
+
+@Plugin(name = "HeaderWarningAppender", category = Core.CATEGORY_NAME, elementType = Appender.ELEMENT_TYPE)
```
I haven't changed `HeaderWarning` though, I've added an appender on top of it. I see the argument that's it's not a 'thing' to write to, but another way of looking at appenders is that they're just 'sinks' for log events. log4j ships with appenders that write to message queues, for example.

We can still add headers using `HeaderWarning`, or configure the appender for other messages as well.

pgomulka replied:
I wonder maybe if we decide to keep `HeaderWarningAppender` then maybe we should create a logger for it and limit the scope of `HeaderWarning`? To make sure that there is single and consistent way of emitting header warnings.

pgomulka replied:
@pugnascotia any views on creating a logger?

pugnascotia replied:
@pgomulka I'm not clear what you are suggesting - are you thinking of adding something like `DeprecationLogger` but for header warnings?

pgomulka replied:
I was thinking of configuring a logger that uses a `header_warning` appender. Named maybe `header_logger` ? and the usage in https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java#L483
would be something like
```
logger = LogManger.getLogger("header_logger");
logger.warn(warning);
```

It worries me that there are two ways to emit those warnings - by log4j and directly.

pugnascotia replied:
Personally, I don't find the current implementation too troubling, but then I was thinking about what the code looks like when we have more places that want to emit warnings back to the client, and even possibly record those to an index. In that world, it starts to make a lot of sense to separate the loggers, so I think I will do as you suggest.

pugnascotia replied:
@pgomulka I've implemented a `header_logger` in 76993c95f96.

rjernst replied:
I'm failing to understand what the *advantage* is of using log4j for header warnings? In either case, the warnings are stored in the thread context still right? Is it for formatting?

If we are going to use it, then I think it should be hidden, like deprecation logging is behind DeprecationLogger. But with header warnings, we already have the HeaderWarning class, and the logger retrieved will be the same for all locations, so we could just do this statically? Using a logger for header warnings directly is odd because the log level doesn't actual matter?

pugnascotia commented on server/src/test/java/org/elasticsearch/common/logging/DeprecationLoggerTests.java:
```diff
@@ -20,72 +20,11 @@
 
 public class DeprecationLoggerTests extends ESTestCase {
-    @SuppressLoggerChecks(reason = "Safe as this is using mockito")
-    public void testLogPermissions() {
```
I removed this test because `DeprecationLogger` is just calling into log4j now.

pugnascotia commented on x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java:
```diff
@@ -0,0 +1,129 @@
+    @Override
+    protected void doStart() {
+        this.appender.start();
+        Loggers.addAppender(LogManager.getLogger("org.elasticsearch.deprecation"), this.appender);
```
I'm not sure I understand the question. The deprecation logger isn't changes, as far as I now, it just has an appender added to it?

pgomulka replied:
My question was if say someone creates `org.elasticsearch.deprecation.somepackageA` before that appender is added to a `org.elasticsearch.deprecation` logger will the child loggers be updated too?
I can see now that `Loggers.addAppender` does a config reload, but might be worth verifying.

pugnascotia commented on qa/evil-tests/src/test/java/org/elasticsearch/common/logging/EvilLoggerTests.java:
```diff
@@ -191,49 +192,6 @@ public void testConcurrentDeprecationLogger() throws IOException, UserException,
 
     }
 
-    public void testDeprecationLoggerMaybeLog() throws IOException, UserException {
```
I removed this test because the rate-limiting functionality now lives in a dedicated Filter.

pugnascotia committed:
Test fixes

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
WIP - trying to write an integration test

pugnascotia committed:
Implement indexing test

pugnascotia committed:
WIP - trying to rewrite DeprecationHttpIT as a rest test

pugnascotia committed:
Tests execute against custom plugin, but need fixed

pugnascotia committed:
Rework deprecation HTTP tests

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2-tests

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
Tweaks

pugnascotia committed:
Merge branch 'index-deprecation-logs-v2-tests' into index-deprecation-logs-v2

pugnascotia committed:
Fix tests finally, by resetting the rate limiting filter

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
Add test

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
Fix license checks

pugnascotia commented:
@rjernst @jakelandis I've now reworked the `DeprecationHttpIT` tests to account for the changes in this PR. Can I ask for a fresh look on this PR, please?

pugnascotia commented on x-pack/plugin/deprecation/qa/rest/src/main/java/org/elasticsearch/xpack/deprecation/TestDeprecationHeaderRestAction.java:
```diff
@@ -90,12 +92,19 @@ public RestChannelConsumer prepareRequest(RestRequest request, NodeClient client
 
+        // Pull out the settings values here in order to guarantee that any deprecation messages are triggered
+        // in the same thread context.
+        final Map<String, Object> settingsMap = new HashMap<>();
```
For some reason, I was tripping over an issue where deprecations in the thread context seemed to get lost during processing. In the end I attributed it to #19222 and reworked the code.

pugnascotia commented on x-pack/plugin/deprecation/qa/rest/src/test/java/org/elasticsearch/xpack/deprecation/DeprecationHttpIT.java:
```diff
@@ -0,0 +1,308 @@
+/**
+ * Tests {@code DeprecationLogger} uses the {@code ThreadContext} to add response headers.
+ */
+public class DeprecationHttpIT extends ESRestTestCase {
```
GitHub lies. This isn't a new file, just...edited a bit.

pugnascotia committed:
Formatting

pugnascotia committed:
Remove redundant line

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into index-deprecation-logs-v2

pgomulka commented on test/framework/src/main/java/org/elasticsearch/test/ESSingleNodeTestCase.java:
```diff
@@ -32,6 +32,7 @@
 import org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings;
 import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.logging.LogConfigurator;
```
unused import?

pgomulka reviewed (approved):
it looks really good now. Thank you for working on this.
I left two minor comments.

pugnascotia committed:
Remove unused import

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into index-deprecation-logs-v2

pugnascotia committed:
Add warning headers solely through the log4j appender

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into index-deprecation-logs-v2

pugnascotia committed:
Fixes to JsonLoggerTest

pugnascotia commented:
@rjernst @jakelandis CI is now green, can you please take another look?

rjernst reviewed (commented):
I left a couple comments, but overall, would it be possible to split this up? The change pretty large, and doing several things, which I think are semi unrelated concerns:

* Using log4j as backend of deprecation logs
* Using log4j as backend of header warnings
* Adding deprecation log messages from log4j to a data stream

pugnascotia commented:
I'm breaking up the PR as follows:

   * [ ] Deprecation logging through log4j #61474
   * [ ] Index deprecation logs

Regarding header warnings: although I thought carefully before reimplementing header warnings with log4j...now I'm not so sure. That part of the work was relatively straightforward, so now I'm thinking that we only go down that road if/when we think we really need to.

pugnascotia commented:
Closing in favour of #61474 and #61484.
